Demo-Android/Gutenberg/src/main/java/org/wordpress/gutenberg/GutenbergView.kt

```diff
@@ -99,38 +91,18 @@ class GutenbergView : WebView {
             }
 
             override fun shouldInterceptRequest(
-                view: WebView?,
-                request: WebResourceRequest?
+                view: WebView,
+                request: WebResourceRequest
             ): WebResourceResponse? {
-                if (request?.url == null) {
-                    return super.shouldInterceptRequest(view, request)
-                } else if(request.url.host?.contains("appassets.androidplatform.net") == true) {
+                val requestResponse = requestInterceptor.modifyRequest(view, request)
```

Comment:
The old design made it impossible for an mis-configured interceptor to cause the request to be run more than once – this change loses that.

I really love the idea of having the Interceptor perform the HTTP request, that part is great, but I wonder if we can get the simplicity back?

Perhaps the interceptor could have another method like `canIntercept(request) -> Bool`?

WDYT?

Reply:
Unfortunately, I do not follow how the proposed changes introduce the possibility of running a request more than once. You may need to elaborate further for me. Are you able to provide an example? 

I'm not opposed to reinstating the protection you describe, but it is difficult for me to imagine how an additional `canIntercept` method might achieve that. Presumably, one could also misconfigure _that_ method and achieve the same duplicative runs. 

I hoisted the performing the HTTP request to the host app because I was unable to mutate what appeared to be read-only request headers in Java (maybe I overlooked something). Additionally, hoisting the networking logic to the host app also provides additional flexibility that the host app needs—e.g., mapping the resource URL to a proxy for private Atomic sites.

Reply:
Considering it further, my latest interpretation is that you mean the error handling that once resided in GutenbergKit ensured that the WebView completed the request if the host app's request interception failed. I don't consider that preventing duplicate runs, but an attempt to provide a fallback. Let me know if I am misinterpreting the situation. 

That said, my understanding is that the `WebResourceRequest` headers and URL cannot be mutated in the host app. Your original example in the description of https://github.com/wordpress-mobile/GutenbergKit/pull/30 appears to showcase it is possible, but I'm experiencing that it is not in the host app's Java file. Maybe there is a limitation in Java that is not present in Kotlin? 

If retaining the previous structure is important, I suppose the host app could manually build a new `WebResourceRequest` and return that?

Reply:
> the error handling that once resided in GutenbergKit ensured that the WebView completed the request if the host app's request interception failed

Yeah, if `requestInterceptor.modifyRequest(request)` returns null, the request will be run again in the `WebView` (and almost certainly fail).

My thinking is that we could:

1. Have something that should return pretty much instantly to say "can we handle this specific request" then
2. Have something that tries the actual request (and can only return a response – there'd be no way to return `null`)

Reply:
> 1. Have something that should return pretty much instantly to say "can we handle this specific request" then

Rather than a separate hook, I used the return value of `null` to achieve this, similar to how [`shouldInterceptRequest`](https://developer.android.com/reference/android/webkit/WebViewClient#shouldInterceptRequest(android.webkit.WebView,%20android.webkit.WebResourceRequest)) functions. Very much like my [previous mirroring](https://github.com/wordpress-mobile/GutenbergKit/pull/34#discussion_r1813399339) of its method signature. 

If you'd prefer to split the two functions—"can I intercept this request?" and "modify this request"—that is fine, I will refactor the code. 

I will share that I do not necessarily perceive the value in doing so. The original implementation still [returned `null`](https://github.com/wordpress-mobile/GutenbergKit/blob/8ed393e2b0bbc770c430f28f533d3994bc3ee53d/Demo-Android/Gutenberg/src/main/java/org/wordpress/gutenberg/GutenbergView.kt#L128-L132) in the event of a failure "handling" the request, which means the WebView would try the request and likely fail.